lib/connectors/mongodb/connector.rb

```diff
@@ -56,30 +70,166 @@ def initialize(configuration: {})
 
+          options = args[1]
+          found_overall = 0
+
+          # if no overall limit is specified by filtering use -1 to not break ingestion, when no overall limit is specified (found_overall is only increased,
+          # thus can never reach -1)
+          overall_limit = -1
+
+          unless options.nil?
+            # there could be a skip parameter defined for filtering
+            skip = Utility::Commons.return_if_present(options[:skip], 0)
+            # there could be a limit parameter defined for filtering -> used for an overall limit (not a page limit, which was introduced for memory optimization)
+            overall_limit = Utility::Commons.return_if_present(options[:limit], -1)
+          end
+
+          overall_limit_reached = false
+
           loop do
-            found_count = 0
+            found_in_page = 0
+
             view = cursor.skip(skip).limit(PAGE_SIZE)
             view.each do |document|
-              yield serialize(document)
-              found_count += 1
+              yield serialize(document) { on_doc_serialization.call if block_given? }
+
+              found_in_page += 1
+              found_overall += 1
+
+              overall_limit_reached = found_overall == overall_limit
+
+              break if overall_limit_reached
             end
 
-            break if found_count == 0
+            page_was_empty = found_in_page == 0
+
+            break if page_was_empty || overall_limit_reached
+
             skip += PAGE_SIZE
           end
         end
       end
 
-      private
+      def setup_aggregate(advanced_config)
+        aggregate = advanced_config[:aggregate]
+
+        pipeline = aggregate[:pipeline]
+        options = extract_options(aggregate)
+
+        if !pipeline_present?(pipeline) && !options_present?(options)
+          Utility::Logger.warn('\'Aggregate\' was specified with an empty pipeline and empty options.')
+        end
+
+        arguments = [pipeline, options]
+
+        [AGGREGATE, arguments]
+      end
+
+      def setup_find(advanced_config)
+        find = advanced_config[:find]
+
+        filter = find[:filter]
+        options = extract_options(find)
+
+        if !filter_present?(filter) && !options_present?(options)
+          Utility::Logger.warn('\'Find\' was specified with an empty filter and empty options.')
+        end
+
+        arguments = [filter, options]
+
+        [FIND, arguments]
+      end
+
+      def extract_options(advanced_config)
+        options_present?(advanced_config[:options]) ? advanced_config[:options] : EMPTY_OPTIONS
+      end
+
+      def find_present?(advanced_config)
+        advanced_config[:find].present?
+      end
+
+      def filter_present?(filter)
+        !filter.nil? && filter != EMPTY_FILTER
+      end
+
+      def aggregate_present?(advanced_config)
+        advanced_config[:aggregate].present?
+      end
+
+      def pipeline_present?(pipeline)
+        !pipeline.nil? && pipeline != EMPTY_PIPELINE
+      end
```

Comment:
1 usage of function found, can be inlined

Reply:
done 👍